dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
+// There's really not that much "logic" here, it's just boilerplate that can be eliminated!
+
+
+// New private scope -- a lot of this is boilerplate for class structures, taken from coffeescript
+(function() {
+  var
+    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
+    __hasProp = {}.hasOwnProperty,
+    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
+    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };
+
+  // class UIComponent: this is a lightweight inheritable class modeled after React
+  UIComponent = (function() {
+
+    function UIComponent($el, props) {
+      this.onClick = __bind(this.onClick, this);
+      this.onUpdate = __bind(this.onUpdate, this);
+      this.onDestroy = __bind(this.onDestroy, this);
+
+      this.$el = $el;
+      this.$el.on('click', this.onClick);
+      this.$el.on('destroyed', this.onDestroy);
+
+      this.props = props || $el.data();
+      this.isMounting = true;
+
+      // Set the initial state
+      this.replaceState( this.getInitialState() || {} );
+    }
+
+    // Destroy a UI component
+    UIComponent.prototype.onDestroy = function() {
+      // Don't do anything if already not mounted
+      if (!this.isMounted())
+        return;
+
+      this.componentWillUnmount();
+
+      // Remove any event handlers
+      var $el = this.$el;
+      this.$el = null;
+      $el.off('click', this.onClick);
+      $el.on('destroyed', this.onDestroy);
+
+      this.componentDidUnmount();
+    }
+
+    // Return a default state for this component
+    UIComponent.prototype.getInitialState = function() { }
+
+    // Extend/merge my state and possibly re-render
+    UIComponent.prototype.setState = function(state) {
+      if (!this.isUpdating) {
+        // Allow things to settle - no need to call more than once per execution
+        this.isUpdating = setTimeout( this.onUpdate, 1 );
+      }
+
+      // Merge this into our state
+      for (var key in state) {
+        if (__hasProp.call(state, key))
+          this.state[key] = state[key];
+      }
+    }
+
+    // Replace my state and possibly re-render
+    UIComponent.prototype.replaceState = function(state) {
+      if (!this.isUpdating) {
+        // Allow things to settle - no need to call more than once per execution
+        this.isUpdating = setTimeout( this.onUpdate, 1 );
+      }
+
+      this.state = state || {};
+    }
+
+    // My state has been changed, and we've let all other processes yield, so we know
+    // there are no more changes coming.  Time to update the DOM.
+    UIComponent.prototype.onUpdate = function(ev) {
```

Comment:
Ah never mind, I see that you expect this to be overridden by the actual impl.  Makes sense to keep this for code completion.